### PR TITLE
Make const getters for Node defs consistent

### PR DESCRIPTION
**Pull request by skottmckay**
Use ConstPointerContainer for Node::ImplicitInputDefs() to be consistent with Node::InputDefs() and Node::OutputDefs().

### CODE REVIEW

File: include/onnxruntime/core/common/const_pointer_container.h [hunk 1]
```diff
@@ -64,6 +64,10 @@ class ConstPointerContainer {
   explicit ConstPointerContainer(const Container& data) noexcept : data_(data) {}
 
   size_t size() const noexcept { return data_.size(); }
+  bool empty() const noexcept { return data_.empty(); }
+
+  ConstIterator cbegin() const noexcept { return ConstIterator(data_.cbegin()); }
+  ConstIterator cend() const noexcept { return ConstIterator(data_.cend()); }
 
   ConstIterator begin() const noexcept { return ConstIterator(data_.cbegin()); }
```
Comment (kkaranasos) [inline include/onnxruntime/core/common/const_pointer_container.h]: Do we need this to be calling data_.cbegin()/cend(), given that we have the cbegin/cend methods now?

File: include/onnxruntime/core/common/const_pointer_container.h [hunk 2]
```diff
@@ -64,6 +64,10 @@ class ConstPointerContainer {
   explicit ConstPointerContainer(const Container& data) noexcept : data_(data) {}
 
   size_t size() const noexcept { return data_.size(); }
+  bool empty() const noexcept { return data_.empty(); }
+
+  ConstIterator cbegin() const noexcept { return ConstIterator(data_.cbegin()); }
```
Comment (linkerzhang) [inline include/onnxruntime/core/common/const_pointer_container.h]: are these two needed?

Comment (skottmckay) [reply include/onnxruntime/core/common/const_pointer_container.h]: > are these two needed?

They make it more consistent with the std Container interfaces. See the Member Function Table here: https://en.cppreference.com/w/cpp/container

File: onnxruntime/core/framework/session_state_initializer.cc [hunk 3]
```diff
@@ -70,13 +72,15 @@ common::Status SessionStateInitializer::CreatePlan(const Node* parent_node,
...
   // ignore any outer scope args we don't know about. this can happen if a node contains multiple subgraphs.
   std::vector<const NodeArg*> valid_outer_scope_node_args;
-  std::for_each(outer_scope_node_args.cbegin(), outer_scope_node_args.cend(),
-                [&mlvalue_name_idx_map, &valid_outer_scope_node_args](const NodeArg* node_arg) {
-                  int idx;
-                  if (mlvalue_name_idx_map.GetIdx(node_arg->Name(), idx).IsOK()) {
-                    valid_outer_scope_node_args.push_back(node_arg);
-                  };
-                });
+  if (outer_scope_node_args) {
```
Comment (kkaranasos) [inline onnxruntime/core/framework/session_state_initializer.cc]: Can this be null?

File: onnxruntime/core/framework/op_kernel_context_internal.h [hunk 4]
```diff
@@ -19,7 +19,7 @@ class OpKernelContextInternal : public OpKernelContext {
                                    IExecutionFrame& frame,
                                    const OpKernel& kernel,
                                    const logging::Logger& logger,
-                                   const std::vector<NodeArg*>& implicit_inputs,
+                                   const ConstPointerContainer<std::vector<NodeArg*>> implicit_inputs,
```
Comment (kkaranasos) [inline onnxruntime/core/framework/op_kernel_context_internal.h]: Should this be a reference or a pointer instead?

File: include/onnxruntime/core/common/const_pointer_container.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/common/const_pointer_container.h]: Given this wrapper is just providing constness, it doesn't matter which one you call. This matches behaviour for a const std::vector where you get std::vector::const_vector regardless of whether you call begin() or cbegin(). 

In general, prefer using cbegin/cend unless you're going to modify something via the iterator as that advertises the intent more clearly, so adding cbegin and cend supports that pattern.

File: onnxruntime/core/framework/op_kernel_context_internal.h (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/framework/op_kernel_context_internal.h]: No. Think of ConstPointerContainer more like an iterator. It's a temporary lightweight wrapper of the container and the caller doesn't need to maintain ownership of it.

**Pull request merged**